File: package/npm/lib/platform.js
"use strict";

const os = require("node:os");
const path = require("node:path");

function target() {
  const goos = {
    darwin: "darwin",
    linux: "linux",
    win32: "windows"
  }[process.platform];
  const goarch = {
    x64: "amd64",
    arm64: "arm64"
  }[process.arch];

  if (!goos || !goarch) {
    throw new Error(`unsupported platform: ${process.platform}/${process.arch}`);
  }

  return { goos, goarch, ext: goos === "windows" ? ".exe" : "" };
}

function packageVersion() {
  return require("../package.json").version;
}

function releaseVersion() {
  return (process.env.SCAFLD_INSTALL_VERSION || packageVersion()).replace(/^v/, "");
}

function releaseTag() {
  return `v${releaseVersion()}`;
}

function assetName(version = releaseVersion()) {
  const t = target();
  return `scafld_${version.replace(/^v/, "")}_${t.goos}_${t.goarch}${t.ext}`;
}

function binaryPath() {
  return path.join(__dirname, "..", "vendor", `scafld${target().ext}`);
}

function repo() {
  return process.env.SCAFLD_GITHUB_REPOSITORY || "nilstate/scafld";
}

function downloadURL() {
  const base = process.env.SCAFLD_INSTALL_BASE_URL;
  if (base) {
    return `${base.replace(/\/$/, "")}/${assetName()}`;
  }
  return `https://github.com/${repo()}/releases/download/${releaseTag()}/${assetName()}`;
}

module.exports = { assetName, binaryPath, downloadURL, releaseTag, releaseVersion, target };

File: package/pypi/src/scafld_launcher/platform.py
import os
import platform
from dataclasses import dataclass


@dataclass(frozen=True)
class Target:
    goos: str
    goarch: str
    ext: str = ""


def target() -> Target:
    system = platform.system().lower()
    machine = platform.machine().lower()

    goos = {
        "darwin": "darwin",
        "linux": "linux",
        "windows": "windows",
    }.get(system)
    goarch = {
        "x86_64": "amd64",
        "amd64": "amd64",
        "aarch64": "arm64",
        "arm64": "arm64",
    }.get(machine)

    if not goos or not goarch:
        raise RuntimeError(f"unsupported platform: {system}/{machine}")
    return Target(goos=goos, goarch=goarch, ext=".exe" if goos == "windows" else "")


def cache_root() -> str:
    if override := os.environ.get("SCAFLD_INSTALL_DIR"):
        return override
    if platform.system().lower() == "windows":
        base = os.environ.get("LOCALAPPDATA") or os.path.expanduser("~\\AppData\\Local")
        return os.path.join(base, "scafld", "bin")
    return os.path.join(os.path.expanduser("~"), ".cache", "scafld", "bin")
